### ISSUE TITLE
Support for multiple covarying GPs

### ISSUE THREAD

**Issue by Bonnevie** [GPflow/GPflow#312]
A number of tasks like co-kriging, derivative observations, and Bayesian quadrature involve two (or more) GPs with different kernels. Often, there's some covariance structure between observations from different GPs in these settings (e.g. integrals and derivatives depend naturally on the function). 

The two GPs can be collected in a single Gaussian field with a block covariance matrix structure so that if we have observations x_1 ~ GP(0, K_11) and x_2 ~ GP(0, K_22)
```
k( (x_1, x_2), (x_1, x_2) )  = [[K_11, K_12], 
                                [K_21, K_22]]
```

which doesn't seem to be too far from something you could model in GPflow. I was thinking that maybe an approach like what's used in the coregional kernel could be repurposed? The framework could maybe also work by defining separate GPs and then have a function that takes the two GPs along with a covariance between them. 

Am I wrong in stipulating that this would be a pretty core thing to support? Derivative observations in particular should be plenty of reason since it integrates rather simply with all existing approaches.

**Comment by jameshensman** [comment]
Hi @Bonnevie , 

Sure, multiple-output GPs fit into the GPflow framework in the way that the Coregion kernel works, exactly as you noted. The idea is that instead of specifying the blocks separately, you simply specify a kernel which takes an extra input column to specify where the blocks are. 

Thus, you can identify any multiple-output GP with a single-output GP, with an  extra indicator on the input. 

The neat thing about this approach is that one only has to implement a new kernel, and then it can plug in to any of the model classes in GPflow to do sparse GPs or VB or MCMC etc. 

I'm not sure that the working model you propose is the way to go, since it's not clear how to ensure that the resulting covariance matrix is positive-definite. Specifying an arbitrary off-diagonal block might lead to non-positive-definiteness. 


As for whether it's a core thing, I'm not sure. I think derivative GPs might be a core thing, so long as the implementation was pretty general. 

Perhaps the first thing to have a go at is a derivative kernel. Take the RBF as an example, and introduce extra columns that specify the derivative(s). That's something that I think many people would be interested in any could form a part of GPflow.

**Comment by Bonnevie** [comment]
Won't you need to exploit the block structure for it to scale well? Of course, using the indicator approach, you could possibly just run a sort prior to computation, but I don't know whether that's a waste of computational effort.

You're right that specifying an arbitrary off-diagonal block is a recipe for disaster. I was thinking of just passing a proper kernel function, but that's not good enough either, I think (imagine an off-diagonal block with extremely high variance coefficient).

Mathematically, the key is that the multiple GPs are often linear transformations of each other which naturally induces a valid covariance, but I'm not sure how that can be exploited.  
 
As a high-level argument for why I think it's a core thing: while I agree that GPflow shouldn't be bloated with functionality, it should (ideally, IMHO) support any natural extensions by third parties so that they can easily implement more complicated models in GPflow inside their own packages and scripts. But of course, this depends on whether you think of GPflow as GPy-in-Tensorflow or Tensorflow-for-GPs. 

I'd love to take a look at a derivative kernel, but I think it would be a bit of an academic detour for me at the moment unfortunately. I'm more likely to need something like co-kriging at the moment (i.e. GPs related by standard affine maps). But I'll come back to the issue when I find the time, unless somebody else picks it up.

**Comment by jameshensman** [comment]
Hi Rasmus,

Re block structure: you'll get a speed up if the matrix can be written as a
Kronecker product. That *is* something I want to see in GPflow, most
certainly. @c0g is also interested in Kronecker structures. You can read a
paper about kronecker structured GPs here:
http://machinelearning.wustl.edu/mlpapers/paper_files/NIPS2011_0443.pdf

There is some code in vff (github.com/jameshensman/vff) that could be
useful.

To do co-krigging, you can actually do a lot with the existing kernels. I
suggest reading this excellent review: https://arxiv.org/abs/1106.6251. The
models discussed can be made using the coregion kernel and taking the
product with another kernel (using the active_dims feature).

To save computation in cokrigging (where the observations are aligned: i.e.
you see every 'output' value at every input value) you can use the
Kronecker stuff, where one of the kernels is the coregion kernel.

**Comment by alexisboukouvalas** [comment]
This is related to the discussion in #79 which perhaps should be closed now.
I am quite interested in taking advantage of the kronecker structure and scaling the methods to many outputs (thousands).
Happy to help on this.

**Comment by mathDR** [comment]
Alexis, I am also interested in this.  Do you know if there is any work
started along this path?  I know in Eigen 3.3, there is (unsupported)
sparse Kronecker products.  Is this the functionality that you would want?
Or would you think homegrown lapack/blas backend is the way to go?

**Comment by jameshensman** [comment]
@mathDR : the first rule of Kronecker products is that one never computes the Kronecker product. The second rule of Kronecker products is, one finds a way to avoid computing the Kronecker product :)

The most important thing to note is that this:

```python
np.dot(np.kron(A, B), x)
```

is the same as this:
```python
X = np.reshape(x, [B.shape[0] , A.shape[0]], order='F')
np.reshape(np.dot(np.dot(B, X), A.T), [-1, 1], order='F')
```

but the latter is *much* cheaper. You can usually do the latter product quickly even when the result of kron(A, B) doesn't fit in memory. 

Check out my vff code, which does Kronecker operations efficiently in pure tensorflow, including operations like matrix_triangular_solve.

**Comment by mathDR** [comment]
Nice.  I will absolutely take a look.  As an aside, do you know if anyone has implemented a spectral mixture kernel yet?  If not, I could also take a stab at that.

**Comment by Bonnevie** [comment]
@mathDR: I for one would be grateful to see spectral mixture kernels in GPflow! Did you see the generalized spectral kernel paper? (https://paperpile.com/shared/Io1WIL)  

@jameshensman: thanks for pointing me towards the vector-valued GP paper. I think I skimmed it a while back, but I'll look into it a bit more deeply now. Now that you have mentioned it a couple of times, does the VFF package add a lot of functionality to GPflow aside from the variational fourier features?

**Comment by jameshensman** [comment]
Recipe for a spectral kernel:

```python
rbf = GPflow.kernels.RBF
cos = GPflow.kernels.Cosine

k = rbf(1) * cos(1) + rbf(1) * cos(1) + rbf(1) * cos(1)
```
simple, right?

VFF doesn't add much other that what I've stated. VFF has to use Kronecker structures, so I worte quite a bit of code to help them along (working from a template by Chris Lloyd). VFF is limited in many ways: you can't use an arbitrary kernel, for example. But for the cases where it does work, it is extremely computationally efficient.

**Comment by alexggmatthews** [comment]
I would like to see Kronecker functionality in GPflow too.

**Comment by mathDR** [comment]
@jameshensman I agree that SM kernels via add/products of rbf and cosine are easily accomplished.  I think the "work" will be in the initialization of the hyperparameters for these.  A lot of the work shows how sensitive good results are on the initialization of these kernels.  That said, perhaps a new thread discussing hyper-priors on hyperparameters?

**Comment by Bonnevie** [comment]
@jameshensman: that construction of the spectral mixture kernel really is beautifully simple and a demonstration of why object-oriented code is great. I had clearly over-dramatized how complicated the kernel is in my head :)
I think the generalized version (at least the non-stationary one) requires a bit more work, but not a lot. 

@mathDR: out of curiosity, what papers show initialization sensitivity? Not that I'm surprised. What worries me is that the choice of kernel is how you encode prior knowledge, so an arbitrarily complex kernel like the SM kernel can model a lot more, yes, but it is also increasingly prone to overfitting.

**Comment by mathDR** [comment]
@Bonnevie Most papers,starting with Wilson's [thesis](https://people.orie.cornell.edu/andrew/#thesis) speak to it.  When Wilson, et.al implemented their kernel in GPML, there was an initialization script to go along with it (that now has an [advanced](https://people.orie.cornell.edu/andrew/code/initSMhypersadvanced.m) version that uses the Fourier series of the observations)  That, and this [comment](https://github.com/SheffieldML/GPy/issues/401#issuecomment-245174035) by Neil Lawrence (when implemented in GPy)

**Comment by Bonnevie** [comment]
@mathDR: uff, that init script is just a recipe for overfitting isn't it? the degenerate kernel yy^T (empirical covariance) is the optimum choice of covariance, so I cannot imagine that fitting the spectral density instead much improves the issue.

**Comment by alexggmatthews** [comment]
I tried the SM kernel with Andrew when he was a student here in Cambridge. Certainly it can overfit like any other flexible parametric model without regularization. The optimization surface can also be very challenging. On the other hand for large audio time series where the ML estimate was fine we found that it worked extremely well.

**Comment by Bonnevie** [comment]
@alexggmatthews: yes, it's definitely a pretty powerful kernel. Did you manage to extrapolate well with it? Even though the original paper was on extrapolation, it seems like the only way the model can extrapolate out into the future is by having inordinately long lengthscales - even in the original paper you see a regression towards the prior. (sorry to derail the discussion a bit)

**Comment by alexggmatthews** [comment]
Certainly the Mauna Loa example in the original paper relied on a very long lengthscale to model the linear component of the trend. 

As far as I can remember the successful audio experiments we did were interpolation of a gap in the data done over a non trivial range of time. 

I'm not sure I'd make a general comment about extrapolation though - it depends on the details.

**Comment by Bonnevie** [comment]
I'm kinda thrilled that the following worked. This [Gist](https://gist.github.com/Bonnevie/de7f0c18a46289d2e57cca8b4b9aea12) gives a method for doing recursive co-kriging (where your observations can be from different levels and each new level is the sum of the GP of the previous level plus a new perturbation GP) where you can observe points at individual levels. 
The basic method is to take in the data and then reorder it so that the matrix has a block structure, then use the block structure to build the kernel matrix, and then reorder to get back to the original ordering.

With all that shuffling, I don't think this will scale super-well, but the bottleneck is most likely the reordering steps, so it would be a lot better if we could assume a priori that the data was ordered or even split.  

It must be possible to use the same order-build-reorder approach to construct other inter-domain GPs.

**Comment by jorgeloaiza** [comment]
Does This recipe    

rbf = GPflow.kernels.RBF
cos = GPflow.kernels.Cosine
k = rbf(1) * cos(1) + rbf(1) * cos(1) + rbf(1) * cos(1)

work in GPy  like a Spectral Mixture kernel?

Thanks.

**Comment by markvdw** [comment]
This should be addressed soon.
